### PR TITLE
Initial Implementation of the getBigQueryData api call

### PR DESCRIPTION
**Pull request by bransteitterbr**
This is for issue: https://github.com/episphere/connect/issues/1126

### CODE REVIEW

File: utils/sites.js [hunk 1]
```diff
@@ -596,8 +596,93 @@ const participantDataCorrection = async (req, res) => {
     }
 };
 
+const getBigQueryData = async (req, res) => {
+    const { validateTableAccess, validateFilters, validateFields, getBigQueryData } = require('./bigquery');
+
+    logIPAddress(req);
+    setHeaders(res);
+    
+    if(req.method === 'OPTIONS') return res.status(200).json({code: 200});
```
Comment (Gbarra9) [inline utils/sites.js:605]:
```suggestion
    if (req.method === 'OPTIONS') return res.status(200).json({ code: 200 });
```

File: utils/sites.js [hunk 2]
```diff
@@ -596,8 +596,93 @@ const participantDataCorrection = async (req, res) => {
...
+
+    if(req.method !== 'GET') {
```
Comment (Gbarra9) [inline utils/sites.js:607]:
```suggestion
    if (req.method !== 'GET') {
```

File: utils/sites.js [hunk 3]
```diff
@@ -596,8 +596,93 @@ const participantDataCorrection = async (req, res) => {
...
+    //     return res.status(500).json(getResponseJSON(authorized.message, 500));
+    // }
+
+    // if(!authorized){
+    //     return res.status(401).json(getResponseJSON('Authorization failed!', 401));
+    // }
+    const authorized = {"siteCode": "NORC"};
+
+    if(req.query.table === undefined || !req.query.table) return res.status(400).json(getResponseJSON('Bad request. Table is not defined in query', 400));
+    if(req.query.dataset === undefined || !req.query.dataset) return res.status(400).json(getResponseJSON('Bad request. Dataset is not defined in query', 400));
```
Comment (Gbarra9) [inline utils/sites.js:622]: Code style/format suggestion, same as above.

File: utils/sites.js [hunk 4]
```diff
@@ -596,8 +596,93 @@ const participantDataCorrection = async (req, res) => {
...
+    }
+    // const { APIAuthorization } = require('./shared');
+    // const authorized = await APIAuthorization(req);
+    // if(authorized instanceof Error){
+    //     return res.status(500).json(getResponseJSON(authorized.message, 500));
+    // }
+
+    // if(!authorized){
+    //     return res.status(401).json(getResponseJSON('Authorization failed!', 401));
+    // }
```
Comment (Gbarra9) [inline utils/sites.js:618]: Does this need to be removed?

File: utils/sites.js [hunk 5]
```diff
@@ -596,8 +596,93 @@ const participantDataCorrection = async (req, res) => {
...
+    let error;
+    let responseArray = [];
+    try {
+        responseArray = await getBigQueryData(dataset, table, filters, fields);
+    } catch (e) {
+        error = e.toString();
+        console.log(e);
+    }
+
+    return res.status(error ? 206 : 200).json({code: error ? 206 : 200, results: responseArray});
```
Comment (Gbarra9) [inline utils/sites.js:680]: I noticed the use of a 206 status code (partial content response) for errors. Can you help me understand the reasoning?

**Review by Gbarra9 — APPROVED** [review]
Looks good, left small comments

File: utils/sites.js [hunk 6]
```diff
@@ -596,8 +596,93 @@ const participantDataCorrection = async (req, res) => {
...
+            filters = [JSON.parse(filters)];
+            if (!filters[0].column || !filters[0].operator) {
+                return res.status(400).json(getResponseJSON('Bad request. Invalid Filter', 400));
+            }
+        } catch (e) {
+            return res.status(400).json(getResponseJSON('Bad request. Filters is not an array or JSON object', 400));
+        }
+    }
+
+    if (Array.isArray(filters)) {
```
Comment (anthonypetersen) [inline utils/sites.js:656]: can both of these be combined?

File: utils/sites.js [hunk 7]
```diff
@@ -596,8 +596,93 @@ const participantDataCorrection = async (req, res) => {
...
+    } else if (typeof filters === "string") {
+        try {
+            filters = [JSON.parse(filters)];
+            if (!filters[0].column || !filters[0].operator) {
+                return res.status(400).json(getResponseJSON('Bad request. Invalid Filter', 400));
+            }
+        } catch (e) {
+            return res.status(400).json(getResponseJSON('Bad request. Filters is not an array or JSON object', 400));
+        }
+    }
```
Comment (anthonypetersen) [inline utils/sites.js:654]: else "bad request" ?

File: utils/bigquery.js [hunk 8]
```diff
@@ -171,9 +171,234 @@ const getParticipantTokensByPhoneNumber = async (fullNumber) => {
...
+/**
+ * getColumnsFromTable - retrieves column details from a table or view
+ * 
+ * @param string dataset
+ * @param string table
+ * @returns Object[]
+ */
+async function getColumnsFromTable (dataset, table) {
+  //@todo
+  //Determine if the view_birthday_card view needs to become a real view
```
Comment (anthonypetersen) [inline utils/bigquery.js:183]: is this still a question?

**Review by anthonypetersen — APPROVED** [review]
see comments

File: utils/sites.js (hunk 4 above)
Comment (bransteitterbr) [reply utils/sites.js]: It needed to be reinstated I had it commented out for local testing.

File: utils/sites.js (hunk 5 above)
Comment (bransteitterbr) [reply utils/sites.js]: I changed it to 500 as that would be the appropriate code

File: utils/sites.js (hunk 3 above)
Comment (bransteitterbr) [reply utils/sites.js:621]: Updated

File: utils/sites.js (hunk 2 above)
Comment (bransteitterbr) [reply utils/sites.js:607]: Updated

File: utils/sites.js (hunk 1 above)
Comment (bransteitterbr) [reply utils/sites.js:605]: Updated

**Pull request merged**